docs/zkapps/o1js/foreign-fields.mdx

```diff
@@ -0,0 +1,111 @@
+
+However, these classes and modules are not compatible with the cryptography used in the wider world: `Signature.verify()` doesn't let you verify a signed JWT or email, and `Encryption.decrypt()` won't help you with your WhatsApp messages. That's because these methods use different finite fields than the native Field that was chosen primarily to enable efficient zk proofs.
+
+Here is where foreign fields come in: They let you perform algorithms that connect your zkApp with the outside world of cryptography. Foreign fields come with an efficiency hit compared to the native Field, but the heavily engineered foreign fields are efficient enough to unlock many interesting use cases.
+
+## Basic usage
+
+<!-- TODO replace with link to /ForeignField when it exists -->
+
+What follows is a brief overview of how to use foreign fields. For more details, refer to the [API reference](../o1js-reference/classes/Field) or the doccomments on each method.
+
+The entry point for using foreign fields is the `createForeignField()` function:
+
+```ts
+import { createForeignField } from 'o1js';
+
+class Field17 extends createForeignField(17n) {}
+```
+
+The only parameter that `createForeignField()` takes is the modulus or size of the field. This code example passes in `17n` so that `Field17` allows you to perform arithmetic modulo 17:
+
+```ts
+let x = Field17.from(16);
+x.assertEquals(-1); // 16 = -1 (mod 17)
+x.mul(x).assertEquals(1); // 16 * 16 = 15 * 17 + 1 = 1 (mod 17)
+```
+
+As modulus, any number of up to 259 bits is supported. This means that `ForeignField` can be used for many elliptic curve algorithms (where bit sizes are often just below 256) but not for RSA with its typical bit size of 2048.
+
+Notably, the modulus does not have to be a prime number. For example, you can create a `UInt256` class where the modulus is `2^256`:
+
+```ts
+class UInt256 extends createForeignField(1n << 256n) {}
+
+// and now you can do arithmetic modulo 2^256!
+let a = UInt256.from(1n << 255n);
+let b = UInt256.from((1n << 255n) + 7n);
+a.add(b).assertEquals(7);
+```
+
+The base type that is common to classes created by `createForeignField()` is `ForeignField`:
+
+```ts
+import { ForeignField } from 'o1js';
+
+// ...
+
+let zero: ForeignField = Field17.from(0);
+let alsoZero: ForeignField = UInt256.from(0);
+```
+
+`ForeignField` supports the basic arithmetic operations:
+
+```ts
+x.add(x); // addition
+x.sub(2); // subtraction
+x.neg(); // negation
+x.mul(3); // multiplication
+x.div(x); // division
+x.inv(); // inverse
+```
+
+Note that these operations happen modulo the field size. So, `Field17.from(1).div(2)` gives 9 because `2 * 9 = 18 = 1 (mod 17)`.
```

Comment:
@mitschabaude  what do we mean "operations happen modulo"

can we remove happen? 
```suggestion
These operations modulo the field size. So, `Field17.from(1).div(2)` gives 9 because `2 * 9 = 18 = 1 (mod 17)`.
```

Reply:
That doesn't read right - sentence doesn't have a verb without happen

Reply:
Let's write "wrap around the field size"

Reply:
Or wrap around the modulus

Reply:
I actually like "happen modulo the field size" and understood it clearly. I think it's grammatically correct or at least common usage. If we have to change it I'd suggest "Note that these operations are performed modulo the field size".

Is 2 * 9 the right example for 1 / 2 though?

Reply:
2 * 9 = 1 demonstrates that 9 = 1/2. but might be confusing (on the other hand, people were always confused by the results of division for `Field` so I like explicitly having it in the docs)